### PR TITLE
Handle TaskCancelledException in AnalyzerManager.GetAnalyzerActionsAsync

### PR DESCRIPTION
**Pull request by mavasani**
AnalyzerManager maintains a cache of per-analyzer tasks to compute sessions and compilation analyis scopes. If multiple clients simultaneously ask for the analyzer actions for a given analyzer, then all of them receive the same task. If the thread that created this task is now cancelled, all other clients get a TaskCancelledException with a different cancellation token then theirs and the exception goes unhandled, bringing down VS.

Fix is to catch this task cancelled exception in AnalyzerManager, remove the entry for the the cancelled task in the scope map and then re-attempt computing scope, if our cancellelation has not fired.

@heejaechang @JohnHamby can you please take a look?

### CODE REVIEW

**Comment by heejaechang** [comment]
what you describing seems what LazyAsync is for.

basically LazyAsync is there can be multiple requester which has their own cancellation token, when all of them cancels, LazyAsync cancels its work otherwise, it will finish what it is asked to do.

and it should handle cancellation correctly. take a look at places where we use LazyAsync

**Comment by heejaechang** [comment]
looks like it is exactly where LazyAsync should be used. it is designed for exactly this kind of situation.

**Comment by mavasani** [comment]
@heejaechang  AsyncLazy lies in Workspaces layer though. I'll try to see if it can be moved to the SharedCollections project. If it ends up being too much work then I would like to at least get the current fix in so that we don't get intermittent VS crashes while just opening Roslyn.sln, and later work on a cleaner fix.

**Comment by heejaechang** [comment]
if it is too fix current crash, I am okay checking in, but I think it should move to using AsyncLazy. it has hierarchy, but you should be able to just copy and paste and use it at worst.

**Comment by heejaechang** [comment]
:+1

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 1]
```diff
@@ -61,7 +61,41 @@ internal class AnalyzerManager
...
+                // Task to compute the scope was cancelled.
+                // Clear the entry in scope map for analyzer, so we can attempt a retry.
+                var compilationActionsMap = _compilationScopeMap.GetOrCreateValue(compilation);
+                Task<HostCompilationStartAnalysisScope> cancelledTask;
+                compilationActionsMap.TryRemove(analyzer, out cancelledTask);
+
+                // Don't attempt retry if our cancellation token fired as the retry will also fail.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return new HostCompilationStartAnalysisScope(sessionScope);
```
Comment (JohnHamby) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: Is it necessary to return a non-null value if cancellation has been requested? What will the returned value be used for in this case?

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 2]
```diff
@@ -77,16 +111,34 @@ internal class AnalyzerManager
...
+            {
+                // Task to compute the scope was cancelled.
+                // Clear the entry in scope map for analyzer, so we can attempt a retry.
+                _sessionScopeMap.Remove(analyzer);
 
-            return task;
+                // Don't attempt retry if our cancellation token fired as the retry will also fail.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return new HostSessionStartAnalysisScope();
```
Comment (JohnHamby) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: As above, why is it necessary to create an object if cancellation has actually been requested?

**Comment by JohnHamby** [comment]
I'm also OK with checking this in.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: I'll replace the if statement with cancellationToken.ThrowIfCancellationRequested.

**Comment by jasonmalinowski** [comment]
If you need it, I'll happily move AsyncLazy to the shared collections layer. After 4 years I _think_ we've got it not-too-terribly-wrong. It's really that hard to do well.

**Pull request merged**